### PR TITLE
feat: add rollup plugins and update config to convert CJS module(`randexp.js`) to ESM format on build 

### PR DESCRIPTION
**Pull request by ryoppippi**
This solves `randregex` library for esm.

Before this PR, we saw that warning like below

https://github.com/samchon/typia/actions/runs/9464501210/job/26071971132
```
(!) Unresolved dependencies
https://rollupjs.org/troubleshooting/#warning-treating-module-as-external-dependency
randexp (imported by "src/utils/RandomGenerator/RandomGenerator.ts")
created lib in 5.7s
```

This PR solves this warning

More clear descriptions here:
- Some bundlers like bun are not good at handling cjs
- [@rollup/plugin-commonjs](https://www.npmjs.com/package/@rollup/plugin-commonjs) converts packages written in commonjs to es6 format, so it resolves dependencies nicely when building esm
- randexp.js is written using `module.exports` and this `@rollup/plugin-commonjs` converts this package to esm format
- also, to find the `randexp.js` code, I added [`@rollup/plugin-node-resolve
`](https://github.com/rollup/plugins/tree/master/packages/node-resolve)

I built with this plugin in hand and the bun build went through fine.

I'll paste a part of the new build result
<details><summary>libs/index.mjs</summary>
<p>

```js

function getDefaultExportFromCjs(x) {
    return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, "default") ? x["default"] : x;
}

/* --------------------------------------- */

var randexp = class RandExp {
    constructor(regexp, m) {
        this._setDefaults(regexp);
        if (regexp instanceof RegExp) {
            this.ignoreCase = regexp.ignoreCase;
            this.multiline = regexp.multiline;
            regexp = regexp.source;
        } else if (typeof regexp === "string") {
            this.ignoreCase = m && m.indexOf("i") !== -1;
            this.multiline = m && m.indexOf("m") !== -1;
        } else {
            throw new Error("Expected a regexp or string");
        }
        this.tokens = ret(regexp);
    }
    _setDefaults(regexp) {
        this.max = regexp.max != null ? regexp.max : RandExp.prototype.max != null ? RandExp.prototype.max : 100;
        this.defaultRange = regexp.defaultRange ? regexp.defaultRange : this.defaultRange.clone();
        if (regexp.randInt) {
            this.randInt = regexp.randInt;
        }
    }
    gen() {
        return this._gen(this.tokens, []);
    }
    _gen(token, groups) {
        var stack, str, n, i, l;
        switch (token.type) {
          case types.ROOT:
          case types.GROUP:
            if (token.followedBy || token.notFollowedBy) {
                return "";
            }
            if (token.remember && token.groupNumber === undefined) {
                token.groupNumber = groups.push(null) - 1;
            }
            stack = token.options ? this._randSelect(token.options) : token.stack;
            str = "";
            for (i = 0, l = stack.length; i < l; i++) {
                str += this._gen(stack[i], groups);
            }
            if (token.remember) {
                groups[token.groupNumber] = str;
            }
            return str;

          case types.POSITION:
            return "";

          case types.SET:
            var expandedSet = this._expand(token);
            if (!expandedSet.length) {
                return "";
            }
            return String.fromCharCode(this._randSelect(expandedSet));

          case types.REPETITION:
            n = this.randInt(token.min, token.max === Infinity ? token.min + this.max : token.max);
            str = "";
            for (i = 0; i < n; i++) {
                str += this._gen(token.value, groups);
            }
            return str;

          case types.REFERENCE:
            return groups[token.value - 1] || "";

          case types.CHAR:
            var code = this.ignoreCase && this._randBool() ? this._toOtherCase(token.value) : token.value;
            return String.fromCharCode(code);
        }
    }
    _toOtherCase(code) {
        return code + (97 <= code && code <= 122 ? -32 : 65 <= code && code <= 90 ? 32 : 0);
    }
    _randBool() {
        return !this.randInt(0, 1);
    }
    _randSelect(arr) {
        if (arr instanceof DRange) {
            return arr.index(this.randInt(0, arr.length - 1));
        }
        return arr[this.randInt(0, arr.length - 1)];
    }
    _expand(token) {
        if (token.type === ret.types.CHAR) {
            return new DRange(token.value);
        } else if (token.type === ret.types.RANGE) {
            return new DRange(token.from, token.to);
        } else {
            let drange = new DRange;
            for (let i = 0; i < token.set.length; i++) {
                let subrange = this._expand(token.set[i]);
                drange.add(subrange);
                if (this.ignoreCase) {
                    for (let j = 0; j < subrange.length; j++) {
                        let code = subrange.index(j);
                        let otherCaseCode = this._toOtherCase(code);
                        if (code !== otherCaseCode) {
                            drange.add(otherCaseCode);
                        }
                    }
                }
            }
            if (token.not) {
                return this.defaultRange.clone().subtract(drange);
            } else {
                return this.defaultRange.clone().intersect(drange);
            }
        }
    }
    randInt(a, b) {
        return a + Math.floor(Math.random() * (1 + b - a));
    }
    get defaultRange() {
        return this._range = this._range || new DRange(32, 126);
    }
    set defaultRange(range) {
        this._range = range;
    }
    static randexp(regexp, m) {
        var randexp;
        if (typeof regexp === "string") {
            regexp = new RegExp(regexp, m);
        }
        if (regexp._randexp === undefined) {
            randexp = new RandExp(regexp, m);
            regexp._randexp = randexp;
        } else {
            randexp = regexp._randexp;
            randexp._setDefaults(regexp);
        }
        return randexp.gen();
    }
    static sugar() {
        RegExp.prototype.gen = function() {
            return RandExp.randexp(this);
        };
    }
};

var RandExp = getDefaultExportFromCjs(randexp);
```

</p>
</details>

### CODE REVIEW

**Comment by ryoppippi** [comment]
This is really need for a part of bun runtime support

File: rollup.config.js [hunk 1]
```diff
@@ -1,15 +1,21 @@
...
   output: {
-    dir: "lib",
+    dir: outDir,
     format: "esm",
     entryFileNames: "[name].mjs",
     sourcemap: true,
   },
   plugins: [
+    nodeResolve(),
+    commomnjs(),
```
Comment (samchon) [inline rollup.config.js:18]: `CommonJS`, is it really okay? Isn't the purpose of `rollup` is to building `esm` file?

**Review by samchon — COMMENTED** [review]
Cannot sure that CommonJS is right configuration.

File: rollup.config.js (hunk 1 above)
Comment (ryoppippi) [reply rollup.config.js:18]: @samchon 
OK, I'll write a clear description here

- Some bundlers like bun are not good at handling cjs
- [@rollup/plugin-commonjs](https://www.npmjs.com/package/@rollup/plugin-commonjs) converts packages written in commonjs to es6 format, so it resolves dependencies nicely when building esm
- randexp.js is written using `module.exports` and this `@rollup/plugin-commonjs` converts this package to esm format
- also, to find the `randexp.js` code, I added [`@rollup/plugin-node-resolve
`](https://github.com/rollup/plugins/tree/master/packages/node-resolve)

I built with this plugin in hand and the bun build went through fine.

I'll paste a part of the new build result
<details><summary>libs/index.mjs</summary>
<p>

```js

function getDefaultExportFromCjs(x) {
    return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, "default") ? x["default"] : x;
}

/* --------------------------------------- */

var randexp = class RandExp {
    constructor(regexp, m) {
        this._setDefaults(regexp);
        if (regexp instanceof RegExp) {
            this.ignoreCase = regexp.ignoreCase;
            this.multiline = regexp.multiline;
            regexp = regexp.source;
        } else if (typeof regexp === "string") {
            this.ignoreCase = m && m.indexOf("i") !== -1;
            this.multiline = m && m.indexOf("m") !== -1;
        } else {
            throw new Error("Expected a regexp or string");
        }
        this.tokens = ret(regexp);
    }
    _setDefaults(regexp) {
        this.max = regexp.max != null ? regexp.max : RandExp.prototype.max != null ? RandExp.prototype.max : 100;
        this.defaultRange = regexp.defaultRange ? regexp.defaultRange : this.defaultRange.clone();
        if (regexp.randInt) {
            this.randInt = regexp.randInt;
        }
    }
    gen() {
        return this._gen(this.tokens, []);
    }
    _gen(token, groups) {
        var stack, str, n, i, l;
        switch (token.type) {
          case types.ROOT:
          case types.GROUP:
            if (token.followedBy || token.notFollowedBy) {
                return "";
            }
            if (token.remember && token.groupNumber === undefined) {
                token.groupNumber = groups.push(null) - 1;
            }
            stack = token.options ? this._randSelect(token.options) : token.stack;
            str = "";
            for (i = 0, l = stack.length; i < l; i++) {
                str += this._gen(stack[i], groups);
            }
            if (token.remember) {
                groups[token.groupNumber] = str;
            }
            return str;

          case types.POSITION:
            return "";

          case types.SET:
            var expandedSet = this._expand(token);
            if (!expandedSet.length) {
                return "";
            }
            return String.fromCharCode(this._randSelect(expandedSet));

          case types.REPETITION:
            n = this.randInt(token.min, token.max === Infinity ? token.min + this.max : token.max);
            str = "";
            for (i = 0; i < n; i++) {
                str += this._gen(token.value, groups);
            }
            return str;

          case types.REFERENCE:
            return groups[token.value - 1] || "";

          case types.CHAR:
            var code = this.ignoreCase && this._randBool() ? this._toOtherCase(token.value) : token.value;
            return String.fromCharCode(code);
        }
    }
    _toOtherCase(code) {
        return code + (97 <= code && code <= 122 ? -32 : 65 <= code && code <= 90 ? 32 : 0);
    }
    _randBool() {
        return !this.randInt(0, 1);
    }
    _randSelect(arr) {
        if (arr instanceof DRange) {
            return arr.index(this.randInt(0, arr.length - 1));
        }
        return arr[this.randInt(0, arr.length - 1)];
    }
    _expand(token) {
        if (token.type === ret.types.CHAR) {
            return new DRange(token.value);
        } else if (token.type === ret.types.RANGE) {
            return new DRange(token.from, token.to);
        } else {
            let drange = new DRange;
            for (let i = 0; i < token.set.length; i++) {
                let subrange = this._expand(token.set[i]);
                drange.add(subrange);
                if (this.ignoreCase) {
                    for (let j = 0; j < subrange.length; j++) {
                        let code = subrange.index(j);
                        let otherCaseCode = this._toOtherCase(code);
                        if (code !== otherCaseCode) {
                            drange.add(otherCaseCode);
                        }
                    }
                }
            }
            if (token.not) {
                return this.defaultRange.clone().subtract(drange);
            } else {
                return this.defaultRange.clone().intersect(drange);
            }
        }
    }
    randInt(a, b) {
        return a + Math.floor(Math.random() * (1 + b - a));
    }
    get defaultRange() {
        return this._range = this._range || new DRange(32, 126);
    }
    set defaultRange(range) {
        this._range = range;
    }
    static randexp(regexp, m) {
        var randexp;
        if (typeof regexp === "string") {
            regexp = new RegExp(regexp, m);
        }
        if (regexp._randexp === undefined) {
            randexp = new RandExp(regexp, m);
            regexp._randexp = randexp;
        } else {
            randexp = regexp._randexp;
            randexp._setDefaults(regexp);
        }
        return randexp.gen();
    }
    static sugar() {
        RegExp.prototype.gen = function() {
            return RandExp.randexp(this);
        };
    }
};

var RandExp = getDefaultExportFromCjs(randexp);
```

</p>
</details>

Comment (ryoppippi) [reply rollup.config.js:18]: With this PR and my PRs https://github.com/ryoppippi/unplugin-typia/pull/82 https://github.com/ryoppippi/unplugin-typia/pull/81 , we can support (`Bun runtime`) (no need build step anymore when we use bun)

Comment (ryoppippi) [reply rollup.config.js:18]: If we remove `randexp.js` as a dependency in the future, we can remove these rollup plugins

Comment (ryoppippi) [reply rollup.config.js:18]: I think this is essential for ESM support

Comment (ryoppippi) [reply rollup.config.js:18]: Also, after this change, vite dev server can solve this

**Review by samchon — COMMENTED** [review]
No matter for publishing v6.1 with this PR, but I've something to ask you.

I've already utilized `typia` in ESM mode of Deno with https://esm.sh/ based bundling, and there had not been any problem. So I cannot understand why the `randexp` makes ESM and Bun to be broken. Looking at your CJS configuration, I understand what do you want, but cannot sure it is the right behavior.

Is this PR really required? Why Bun can't do it? If I accept this PR, the ESM module users may actually not utilize the `randexp` library. If the `randexp` performs a minor or patch version update, `typia` ESM module user never can take advantage of such update. This is the one thing I'm worrying about.

> Even though of my curiosity, you say that this PR still be required, I'll publish v6.1 with merging.

**Comment by ryoppippi** [comment]
@samchon 
This is why this PR is needed.
- As can be seen from the `unresolved dependencies` warning, there is no `default` export for `randexp`. This is why some bundlers don't handle it well
- Some bundler including `Bun`'s dev mode or vite using `esbuild` dev mode have errors due to insufficient cjs optimisation(like `require` is not defineded)
- There is a mechanism in esm.sh to convert the default export of cjs (see https://esm.sh/typia@6.0.6/lib/utils/RandomGenerator).
- @rollup/plugin-commonjs solves and bundles libraries written in cjs without defaults when it finds them
- randexp has not been released since 2018, there is an issue standing for ES5 support, but the author is not interested in supporting it

**Comment by samchon** [comment]
Got it, may I publish v6.1.0 update right now? Everything be ready?

**Comment by ryoppippi** [comment]
Yes! It works really fine on my local!

**Comment by ryoppippi** [comment]
@samchon 
Wait I'll update a docs

**Pull request merged**